src/datastore.jl

```diff
@@ -19,6 +40,58 @@ using .Threads
 const datastore = Dict{Int,RefState}()
 const datastore_lock = Ref{Union{Nothing, ReentrantLock}}(nothing)
 const id_counter = Ref{Union{Nothing, Atomic{Int}}}(nothing)
+const datastore_counter = Dict{Tuple{Int,Int}, Atomic{Int}}()
+const local_datastore_counter = Dict{Tuple{Int,Int}, Atomic{Int}}()
+
+function poolref(d::DRef)
+    with_datastore_lock() do
+        ctr = get!(local_datastore_counter, (d.owner,d.id)) do
+            # We've never seen this DRef, so tell the owner
+            if d.owner == myid()
+                poolref_owner(d)
+            else
+                @async remotecall(poolref_owner, d.owner, d)
+            end
+            Atomic{Int}(0)
+        end
+        atomic_add!(ctr, 1)
+    end
```

Comment:
so is this getting the refcount? can local counter go out of sync from the counter stored on owner?

Reply:
The local counter is intended to be strictly local to a given worker, and removes the need to do any network communication with the owner until a worker loses all references to a given `DRef` (and also when it first gets a reference to a `DRef` it's never seen). The non-local counter is the owner's counter of how many workers hold a reference to the `DRef` (including itself). The two are not intended to be synchronized, since that would require lots of network traffic.

Reply:
And no, `poolref` and `poolunref` don't intentionally return a useful value (although they probably return the last value right now). The intent is for MemPool to do all the refcounting by itself, without the user or libraries having to deal with it.